controllers/notebook/notebook_controller.go

```diff
@@ -0,0 +1,506 @@
+	nbIngress := &v1.Ingress{}
+	err = r.Get(context.Background(), nbIngressName, nbIngress)
+	ingressPath := getIngressPath(nbService)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("create notebook ingress " + nbIngressName.String())
+			ingress := generateIngress(nbService, notebook, ingressPath)
+			// create the notebook ingress
+			err = r.Create(context.Background(), ingress)
+			if err != nil {
+				r.Log.Error(err, "failed to create ingress "+nbIngressName.String())
+				return reconcile.Result{Requeue: true}, err
+			}
+			r.recorder.Eventf(notebook, core.EventTypeNormal, "SuccessfulCreateIngress", "Created ingress %v", nbService.Name)
+		} else {
+			// for other errors
+			return reconcile.Result{}, err
+		}
+	}
+
+	if notebook.Status.Url == "" {
+		if len(nbIngress.Status.LoadBalancer.Ingress) > 0 {
+			fullUrl := "http://"
+			ingress := nbIngress.Status.LoadBalancer.Ingress[0]
+			if ingress.IP != "" {
+				fullUrl += ingress.IP + ingressPath + "/"
+			} else if ingress.Hostname != "" {
+				fullUrl += ingress.Hostname + ingressPath + "/"
+			}
+			notebook.Status.Url = fullUrl
+			// update notebook url
+			err = r.Status().Update(context.Background(), notebook)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		} else {
+			shouldReconcile = true
+		}
+	}
+	return ctrl.Result{Requeue: shouldReconcile}, nil
+}
+
+func getNotebookServiceName(notebook *notebookv1alpha1.Notebook) types.NamespacedName {
+	nbServiceName := types.NamespacedName{Namespace: notebook.Namespace, Name: nbNamePrefix(notebook.Name)}
+	return nbServiceName
+}
+
+func (r *NotebookReconciler) updateNotebookCondition(pod *corev1.Pod, notebook *notebookv1alpha1.Notebook) (bool, error) {
+	var err error
+	var shouldReconcile bool
+	if pod.Status.Phase == corev1.PodPending {
+		if notebook.Status.Condition == "" {
+			notebook.Status.Condition = notebookv1alpha1.NotebookCreated
+			notebook.Status.Message = "created notebook pod " + pod.Name
+			notebook.Status.LastTransitionTime = metav1.Now()
+			err = r.Status().Update(context.Background(), notebook)
+		}
+		shouldReconcile = true
+	} else if pod.Status.Phase == corev1.PodRunning {
+		if notebook.Status.Condition != notebookv1alpha1.NotebookRunning {
+			notebook.Status.Condition = notebookv1alpha1.NotebookRunning
+			notebook.Status.LastTransitionTime = metav1.Now()
+			notebook.Status.Message = "notebook pod " + pod.Name + " is running"
+			err = r.Status().Update(context.Background(), notebook)
+		}
+	} else if pod.Status.Phase == corev1.PodFailed || pod.Status.Phase == corev1.PodSucceeded {
+		if notebook.Status.Condition != notebookv1alpha1.NotebookTerminated {
+			notebook.Status.Condition = notebookv1alpha1.NotebookTerminated
+			notebook.Status.LastTransitionTime = metav1.Now()
+			notebook.Status.Message = "notebook pod " + pod.Name + " terminated: " + string(pod.Status.Phase)
+			err = r.Status().Update(context.Background(), notebook)
+		}
+	}
```

Comment:
this could be simplified as below:

```go
var(
  condition, message string
)

switch pod.Status.Phase
case corev1.PodPending:
    condition=...
    message=...
case corev1.PodRunning:
    condition=...
    message=...
case corev1.PodFailed, corev1.PodSucceeded:
    condition=...
    message=...

notebook.Status.Condition = condition
notebook.Status.Message = msg
notebook.Status.LastTransitionTime = metav1.Now()
err = r.Status().Update(context.Background(), notebook)
...
```

Reply:
fixed